### PR TITLE
Recurrences

### PR DESCRIPTION
**Pull request by dbleek**
This PR:

- Adds `is_recurring`, `recurrence` and `end_date `fields to the Opportunity model
- Includes these fields on the post/update opportunity templates, and disables `recurrence` and `end_date` when  `is_recurring=False`
- Raises an error is `is_recurring=True` and  `recurrence=None` during form validation
- Streamlines the Opportunity save for less repetition/easier customization

### CODE REVIEW

File: voluncheer/opportunityboard/views/postanopportunity.py [hunk 1]
```diff
@@ -47,14 +51,18 @@ def update_an_opportunity(request, opportunity_id):
...
-        form = PostAnOpportunityForm(request.POST, request.FILES)
+        form = PostAnOpportunityForm(request.POST, request.FILES, instance=opportunity_to_update)
         if form.is_valid():
             if "delete" in request.POST:
                 form.delete(opportunity_id)
             else:
-                form.update(opportunity_id)
+                form.save()
         else:
-            print(form.errors.as_data())
```
Comment (JinweiZhu) [inline voluncheer/opportunityboard/views/postanopportunity.py:57]: We shouldn't render if the form is invalid. We should throw an error right?

File: voluncheer/opportunityboard/test_forms.py [hunk 2]
```diff
@@ -82,8 +84,28 @@ def test_update(self):
...
+            "end": "12:00:00",
+            "address_1": "New York, NY",
+            "address_2": "Mean st",
+            "is_published": True,
+            "is_recurring": True,
+            "recurrence": None,  # Recurrence set to none despite is_recurring = True
+        }
+
+        form = PostAnOpportunityForm(data=data)
+        self.assertFalse(form.is_valid())
```
Comment (JinweiZhu) [inline voluncheer/opportunityboard/test_forms.py:111]: We should have at least one test that make sure the new fields added above are set to the desired value, no?

File: voluncheer/opportunityboard/models.py [hunk 3]
```diff
@@ -80,6 +85,9 @@ class Meta:
     description = models.TextField()
     date = models.DateTimeField()
     end = models.TimeField()
+    is_recurring = models.BooleanField(default=False)
+    recurrence = models.CharField(max_length=6, choices=FREQUENCIES, null=True, blank=True)
```
Comment (JinweiZhu) [inline voluncheer/opportunityboard/models.py:89]: Are you reusing FREQUENCIES anywhere else?

File: voluncheer/opportunityboard/views/postanopportunity.py [hunk 4]
```diff
@@ -26,7 +26,11 @@ def post_an_opportunity(request):
         if form.is_valid():
             form.save()
         else:
-            print(form.errors.as_data())
```
Comment (JinweiZhu) [inline voluncheer/opportunityboard/views/postanopportunity.py:29]: If the request is a post and form is invalid, then it shouldn't render.

File: voluncheer/opportunityboard/test_forms.py (hunk 2 above)
Comment (dbleek) [reply voluncheer/opportunityboard/test_forms.py:111]: Added fields to test_models.py.

File: voluncheer/opportunityboard/models.py (hunk 3 above)
Comment (dbleek) [reply voluncheer/opportunityboard/models.py:89]: Moved `FREQUENCIES `under Opportunity model definition

File: voluncheer/opportunityboard/views/postanopportunity.py (hunk 1 above)
Comment (dbleek) [reply voluncheer/opportunityboard/views/postanopportunity.py:57]: If we don't render, the user will not see the error message and will not know why the opportunity failed to be created.

File: voluncheer/opportunityboard/views/postanopportunity.py (hunk 4 above)
Comment (dbleek) [reply voluncheer/opportunityboard/views/postanopportunity.py:29]: See above

**Pull request merged**